tests/components/transmission/conftest.py

```diff
@@ -0,0 +1,229 @@
+    CONF_HOST,
+    CONF_LIMIT,
+    CONF_NAME,
+    CONF_ORDER,
+    CONF_PASSWORD,
+    CONF_PORT,
+    CONF_USERNAME,
+    DOMAIN,
+)
+from homeassistant.components.transmission.const import (
+    ORDER_BEST_RATIO_FIRST,
+    ORDER_NEWEST_FIRST,
+    ORDER_OLDEST_FIRST,
+    ORDER_WORST_RATIO_FIRST,
+)
+
+from tests.common import MockConfigEntry, load_fixture
+
+MOCK_TORRENTS_OLDEST_TO_NEWEST = [2, 1, 3, 4, 5]
+MOCK_TORRENTS_BEST_TO_WORST = [3, 1, 5, 2, 4]
+
+
+@pytest.fixture(name="api")
+def mock_transmission_api():
+    """Mock an api."""
+    with patch("transmissionrpc.Client"):
+        yield
+
+
+@pytest.fixture(name="auth_error")
+def mock_api_authentication_error():
+    """Mock an api."""
+    with patch(
+        "transmissionrpc.Client", side_effect=TransmissionError("401: Unauthorized")
+    ):
+        yield
+
+
+@pytest.fixture(name="conn_error")
+def mock_api_connection_error():
+    """Mock an api."""
+    with patch(
+        "transmissionrpc.Client",
+        side_effect=TransmissionError("111: Connection refused"),
+    ):
+        yield
+
+
+@pytest.fixture(name="unknown_error")
+def mock_api_unknown_error():
+    """Mock an api."""
+    with patch("transmissionrpc.Client", side_effect=TransmissionError):
+        yield
+
+
+@pytest.fixture(name="transmission_setup")
+def transmission_setup_fixture():
+    """Mock transmission entry setup."""
+    with patch(
+        "homeassistant.components.transmission.async_setup_entry", return_value=True
+    ):
+        yield
+
+
+def get_correct_fixture(*args):
+    """Return the appropriate JSON string."""
+    method = json.loads(args[0])["method"]
+    if method == "torrent-get":
+        return load_fixture("transmission-torrents-get.json")
+    if method == "session-get":
+        return load_fixture("transmission-session-get.json")
+    if method == "session-stats":
+        return load_fixture("transmission-session-stats.json")
+    return ""
+
+
+def setup_transmission(
+    hass,
+    name="Transmission",
+    host="0.0.0.0",
+    username="user",
+    password="pass",
+    port=9091,
+    limit=20,
+    order=ORDER_OLDEST_FIRST,
+):
+    """Init the mock Transmission entry."""
+    entry = MockConfigEntry(
+        domain=DOMAIN,
+        data={
+            CONF_NAME: name,
+            CONF_HOST: host,
+            CONF_USERNAME: username,
+            CONF_PASSWORD: password,
+            CONF_PORT: port,
+            CONF_LIMIT: limit,
+            CONF_ORDER: order,
+        },
+    )
+    entry.add_to_hass(hass)
+    return entry
+
+
+@pytest.fixture(name="torrent_info")
+async def mock_torrent_info(hass):
+    """Mock Transmission setup with response from the RPC."""
+    entry = setup_transmission(hass)
+    with patch("transmissionrpc.Client._http_query", side_effect=get_correct_fixture):
+        await hass.config_entries.async_setup(entry.entry_id)
+        await hass.async_block_till_done()
+        yield
+
+
+@pytest.fixture(name="status_seeding")
```

Comment:
Many of these fixtures map 1:1 to an existing test case.  I find that it may be tough to understand exactly what the test is doing when reading it because you have to jump back and forth between both the test and the common setup.  (For example, understanding that torrent_limit is truncating two entries required an extra jump back and forth)

The rationale here for doing this makes sense because of the boiler plate setup that you want to hide from the test -- otherwise you have the opposite problem of not knowing what is unique to the particular test case.

Might there be another option for sharing code while still leaving test specific setup within the test itself?  You could have a single client setup method shared for all tests:
```
async def mock_client_setup(hass, entry=None):
    """Mock Transmission client setup."""
    if entry is None:
        entry = setup_transmission(hass)
    with patch("transmissionrpc.Client._http_query", side_effect=get_correct_fixture):
        await hass.config_entries.async_setup(entry.entry_id)
        await hass.async_block_till_done()
        return hass.data[DOMAIN][entry.entry_id]
```

Then the test still has minimal code but you can see right in line the manipulation being done on the client library
```
async def test_status_seeding(hass):
    """Test the Transmission Status sensor."""
    tm_client = await mock_client_setup(hass)
    # pylint: disable=protected-access
    tm_client._tm_data.data.downloadSpeed = 0

    status = hass.states.get("sensor.transmission_status")
    assert status.state == "Seeding"
```

Then for tests that have a special arguments to `setup_transmission`, you can also have that in the test because it also has interesting test-specific, but then pass the returned entry as an argument to `mock_client_setup`.

My suggestion above may not work because the async_block_till_done could need to be done after update the _tm_data, so in that case since there are only two parameters on the client that are modified (uploadSpeed and downloadSpeed) then these could also be options arguments passed into mock_client_setup() which would be fairly readable.

This may be bordering on style preference, just something I thought make review of a lot of test code go faster.  What do you think?

Reply:
@allenporter Thanks so much!  I was struggling with this and ended up with ctrl-c / ctrl-v so I'm glad to learn a better approach.  I will test this out and update the PR if I can get it working.  Since I'm just learning the HA tests I spent a lot of time simply figuring out how to modify the HA entity - once I had it I said "good enough"...

The line `return hass.data[DOMAIN][entry.entry_id]` should make the whole thing do-able and readable while reducing the number of fixtures to a reasonable amount!

Reply:
This worked very well and you were right about the `async_block_till_done`.  I ended up adding a kwargs and setting the attributes in the `mock_client_setup` helper.

```python3
async def mock_client_setup(hass, entry=None, **kwargs):
    """Mock Transmission client setup."""
    # pylint: disable=protected-access
    if entry is None:
        entry = setup_transmission(hass)
    with patch(
        "homeassistant.components.transmission.transmissionrpc.Client._http_query",
        side_effect=get_correct_fixture,
    ):
        await hass.config_entries.async_setup(entry.entry_id)
        for k, v in kwargs.items():
            setattr(hass.data[DOMAIN][entry.entry_id]._tm_data.data, k, v)
        await hass.async_block_till_done()
```
I will understand if @MartinHjelmare or anyone else doesn't approve... :smile: